### PR TITLE
cpu/sam3:update gpio-driver

### PR DESCRIPTION
**Pull request by TobiasFredersdorf**
Updated the gpio peripheral driver implementation for sam3x CPUs.

### CODE REVIEW

**Comment by PeterKietzmann** [comment]
@TobiasFredersdorf I'm on holiday for the next two week. @katezilla (Katja from our last Hack'n'ACK) will probably review your  PR

**Comment by LudwigOrtmann** [comment]
Please add a comment to the commit message why/how/what this update is.

File: boards/arduino-due/include/periph_conf.h [hunk 1]
```diff
@@ -142,207 +142,6 @@ extern "C" {
...
-#define GPIO_30_DEV         PIOC
-#define GPIO_30_PIN         PIO_PC13
-#define GPIO_30_IRQ         PIOC_IRQn
-#define GPIO_C13_MAP        30
-/* GPIO channel 31 config */
-#define GPIO_31_DEV         PIOB
-#define GPIO_31_PIN         PIO_PB21
-#define GPIO_31_IRQ         PIOB_IRQn
-#define GPIO_B21_MAP        31
-/** @} */
```
Comment (katezilla) [inline boards/arduino-due/include/periph_conf.h]: please remove one of the new lines

File: cpu/sam3/include/periph_cpu.h [hunk 2]
```diff
@@ -1,35 +1,79 @@
...
+#ifndef PERIPH_CPU_H_
+#define PERIPH_CPU_H_
+
+#include "periph/dev_enums.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
```
Comment (katezilla) [inline cpu/sam3/include/periph_cpu.h]: two new line here as well

File: cpu/sam3/include/periph_cpu.h [hunk 3]
```diff
@@ -1,35 +1,79 @@
...
+#define HAVE_GPIO_FLANK_T
+typedef enum {
+    GPIO_RISING = 1,        /**< emit interrupt on rising flank */
+    GPIO_FALLING = 2,       /**< emit interrupt on falling flank */
+    GPIO_BOTH = 3           /**< emit interrupt on both flanks */
+} gpio_flank_t;
+/** @} */
+
+/**
+ * @brief   Available ports on the STM32F1 family
```
Comment (katezilla) [inline cpu/sam3/include/periph_cpu.h]: isn't this sam3, not stm32f1?

File: cpu/sam3/include/periph_cpu.h [hunk 4]
```diff
@@ -1,35 +1,79 @@
...
+ */
+enum {
+    PORT_A = 0,             /**< port A */
+    PORT_B = 1,             /**< port B */
+    PORT_C = 2,             /**< port C */
+    PORT_D = 3,             /**< port D */
+    
+};
+
+
```
Comment (katezilla) [inline cpu/sam3/include/periph_cpu.h]: too many new lines

File: cpu/sam3/include/periph_cpu.h [hunk 5]
```diff
@@ -1,35 +1,79 @@
...
+/**
+ * @brief declare needed generic SPI functions
```
Comment (katezilla) [inline cpu/sam3/include/periph_cpu.h]: behind every other brief are spaces, not here. please remove the spaces after the other brief's

**Comment by katezilla** [comment]
congrats on your first pr :-) i will test it in a few days when i have the board

File: cpu/sam3/include/periph_cpu.h [hunk 6]
```diff
@@ -1,35 +1,79 @@
...
-
-/**
- * @ingroup         cpu_sam3x8e
- * @{
- *
- * @file
- * @brief           CPU specific definitions for internal peripheral handling
- *
- * @author          Hauke Petersen <hauke.peterse@fu-berlin.de>
- */
```
Comment (katezilla) [inline cpu/sam3/include/periph_cpu.h]: why did you remove the header?

**Comment by katezilla** [comment]
okay i tried testing a few pins but im not sure if i broke the arduino-due or your code does not work. not all pins thou but quite a few... did you test it with tests/periph_gpio?

**Comment by TobiasFredersdorf** [comment]
The Pins i've  tested,  works. But i will test them again

**Comment by katezilla** [comment]
mind telling me which pins and what you did?

**Comment by katezilla** [comment]
your branch does not even compile... i were able to build because i had at first merge conflilcts from my master and solved them.
but now A-Paul showed me how i take only your branch :-)
please resolve your conflicts before i can go back to testing

**Comment by PeterKietzmann** [comment]
@TobiasFredersdorf will you address @katezilla's comments? If you're ready, please give me a sign so we can test your implementation again.

**Comment by TobiasFredersdorf** [comment]
Now it is up-to-date. And I think it works now. @PeterKietzmann you can test it now

**Comment by haukepetersen** [comment]
Seems like there is a significant problem with external interrupts here: you trigger once callback per port, independent of which pin actually triggered the event...

Let me PR something against your branch, that should fix this...

**Comment by haukepetersen** [comment]
and here it is https://github.com/TobiasFredersdorf/RIOT/pull/1

**Comment by PeterKietzmann** [comment]
@haukepetersen did you test the driver including your changes? I couldn't get the external interrupts running.

**Comment by haukepetersen** [comment]
I did once, but I might have messed up at some point...

**Comment by PeterKietzmann** [comment]
Could you give it a quick try? Otherwise it might take us "hours" to find the reason for the non-working extis.

**Comment by haukepetersen** [comment]
fixed -> https://github.com/TobiasFredersdorf/RIOT/pull/2

also included the clearing of contexts, so now we won't run out of memory when re-defining pin functions...

**Comment by PeterKietzmann** [comment]
@TobiasFredersdorf I see that you merged Haukes latest change into your PR. Can you please check if it works and give me a feedback then

**Comment by TobiasFredersdorf** [comment]
@PeterKietzmann  I will test it and tell you the results

**Comment by TobiasFredersdorf** [comment]
@haukepetersen  Can you look at the code for the interrupt-control because the interrupts for port A are broken. Every time the Port A  triggered an interrupt the micro-controller reboots. I can't find the mistake in the software

**Comment by haukepetersen** [comment]
I see the same behavior here, also simple reading any pin from port A as simple input does not work

**Comment by haukepetersen** [comment]
A quick debug session showed me, that the code for `gpio_init` and `gpio_read` seems to be ok, all the used inline functions give the right addresses/offsets. So there seems to be something special with port A? I don't have the time to look into this further, though...

My next steps would be to scan the datasheet/ref manual thoroughly to see if we missed anything for port A.

**Comment by haukepetersen** [comment]
Sorry for the delay, I found the bug: https://github.com/TobiasFredersdorf/RIOT/pull/3

Could you please squash and rebase, so that we can get this PR in? Thanks!

**Comment by haukepetersen** [comment]
@TobiasFredersdorf ping

**Comment by TobiasFredersdorf** [comment]
i hope it works tomorow. I think i broke the Pull-Request

**Comment by PeterKietzmann** [comment]
@TobiasFredersdorf what do you mean by "broke" ? If you feel like you can't rebase and everything got stuck in your git you could try the following, even if this is not the prettiest way. Please note that this is just a proposal if nothing works any more. But before you should try to rebase or at least know what your problem is. Now the alternative is:

1. Get a new RIOT clone
2. Checkout a new branch
3. Add your changes to the new RIOT
4. Check if everything still works as expected
5. Close this PR and create a new one

**Please** (!!!) read our  [Developement procedures](https://github.com/RIOT-OS/RIOT/wiki/Development-procedures). When it comes to git the [git cheatsheet](https://github.com/RIOT-OS/RIOT/wiki/Git-cheatsheet) could also be of help.

**Comment by haukepetersen** [comment]
Also not quite sure where you got stuck. Maybe just rebase and deleted the lastest commit while doing this? So here is how it worked for me to rebase and fix your branch:
- check it out
- do a `git rebase -i origin/master` (if your up-to-date upstream master is `origin/master`... )
- in the rebase view, delete the line `pick 4e057e1 Update`
- save the file and exit
- now git will complain about merge problems, so you have to do a view iterations of this:
   - fix merge problem by either using some mergetool or manually editing the `cpu/sam3/periph_cpu.h` which git is complaining about
   - add the file `git add cpu/sam3/periph_cpu.h`
   - contiunue rebasing: `git rebase --continue`
   - do this until all commits are proberly rebased
done, your branch should be fixed...

**Comment by PeterKietzmann** [comment]
Can be closed in favour of #4670

**Comment by haukepetersen** [comment]
nope, can not, #4670 does not contain all of the code..

**Comment by haukepetersen** [comment]
my bad, the gpio file was hidden by github... never mind.
